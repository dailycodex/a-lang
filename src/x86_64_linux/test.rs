macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            use super::*;
            use $crate::ir;
            use $crate::lexer::lex;
            use $crate::parse::parse;
            let contents = include_str!($path);
            let tokens = lex(contents).unwrap();
            let ast = parse(tokens).unwrap();
            let ir_code = ir::code_gen(ast).unwrap();
            let asm_data = compile_ir_code(ir_code).unwrap();
            let result = instruction_to_string(asm_data).unwrap();
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(result);
            });
        }
    };
}

snapshot!(binary, "testdata/snapshots/binary.a");
snapshot!(ifelse, "testdata/snapshots/ifelse.a");
