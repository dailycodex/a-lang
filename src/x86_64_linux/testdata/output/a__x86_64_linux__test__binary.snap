---
source: src/x86_64_linux/test.rs
expression: result
---
main__:
      push       rbp
       mov       rbp,       rsp
       mov       rdi,         1
       mov       rsi,         2
       mov       rdx,         3
       mov       rcx,       rsi
      imul       rcx,       rdx
       mov       rsi,       rdi
       add       rsi,       rcx
.exit__:
       mov       rbp,       rsp
       pop       rbp
       ret

